overlord/devicestate/handlers.go

```diff
@@ -624,17 +699,25 @@ func (m *DeviceManager) doRequestSerial(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	finish := func(serial *asserts.Serial) error {
+		if regCtx.FinishRegistration(serial); err != nil {
+			return err
+		}
+		t.SetStatus(state.DoneStatus)
+		return nil
+	}
+
 	if len(serials) == 1 {
 		// means we saved the assertion but didn't get to the end of the task
-		return m.finishRegistration(t, device, serials[0].(*asserts.Serial))
+		return finish(serials[0].(*asserts.Serial))
 	}
 	if len(serials) > 1 {
 		return fmt.Errorf("internal error: multiple serial assertions for the same device key")
 	}
```

Comment:
That's is exiting code, but wondering if len(serials)==0 is possible here, and should lead to internal error too?

Reply:
yes, len = 0 is possible but the query is done not to redo work/be idempotent,  len = 0 is the base case, it's not an error but it's when the task needs to do the complete work